grammar/grammar.py

```diff
@@ -100,355 +97,330 @@ def generate(self, root='root'):
+                        args = popped_scope['args'] + args
+
+                        # execute
+                        func_ret = func_to_exec(scope, frame_stack, user_vars, args)
+                        # treat the character after the paren as our close
+                        i += 1
+                        char = exp[i]
+                        if char == '.':
+                            # pass output to next function if dot
+                            scope.append({'scope_type':'.', 'token':'', 'close_char':popped_scope['close_char'], 'args':[func_ret]})
+                            if 'varstore' in popped_scope:
+                                scope[-1]['varstore'] = popped_scope['varstore']
+                        elif char == popped_scope['close_char'] or char == ',' or char == '.':
                             if popped_scope['close_char'] == '>':
-                                scope[-1]['scope_type'] = '<'
-                            if popped_scope['close_char'] == ']':
-                                scope[-1]['scope_type'] = '['
+                                # tags are not output
+                                if 'tags' not in scope[-1]:
+                                    scope[-1]['tags'] = [func_ret]
+                                else:
+                                    scope[-1]['tags'].append(func_ret)
+                            elif popped_scope['close_char'] != ']':
+                                scope[-1]['token'] += func_ret
+                            if 'varstore' in popped_scope:
+                                user_vars[popped_scope['varstore']] = func_ret
+                            # if close was a comma, chain the invocation scope
+                            if char == ',':
+                                scope.append({'scope_type':popped_scope['close_char'], 'token':''})
+                                if popped_scope['close_char'] == '>':
+                                    scope[-1]['scope_type'] = '<'
+                                if popped_scope['close_char'] == ']':
+                                    scope[-1]['scope_type'] = '['
+                        else:
+                            self.raise_parse_error(f"Function call not at end of invocation. Expression: {exp} function call: {popped_scope['close_char']}{args[0]}.{func_name}({str(args)[1:-1]}){char}", scope, user_vars, frame_stack)
                     else:
-                        self.raise_parse_error(f"Function call not at end of invocation. Expression: {exp} function call: {popped_scope['close_char']}{args[0]}.{func_name}({str(args)[1:-1]}){char}", output, user_vars, parent_stack)
-                else:
-                    scope[-1]['token'] += char
+                        scope[-1]['token'] += char
 
-            i += 1
+                i += 1
+            
+            # inverted do-while condition
+            if not frame_stack:
+                break
```

Comment:
Again, better to ensure the while loop will always run at least once than do this.